lib/libz/zlib.3
.\"	$OpenBSD: zlib.3,v 1.1 2003/09/25 09:12:09 jmc Exp $
.\"
.\"  Copyright (C) 1995-2002 Jean-loup Gailly and Mark Adler
.\"
.\"  This software is provided 'as-is', without any express or implied
.\"  warranty.  In no event will the authors be held liable for any damages
.\"  arising from the use of this software.
.\"
.\"  Permission is granted to anyone to use this software for any purpose,
.\"  including commercial applications, and to alter it and redistribute it
.\"  freely, subject to the following restrictions:
.\"
.\"  The origin of this software must not be misrepresented; you must not
.\"  claim that you wrote the original software. If you use this software
.\"  in a product, an acknowledgment in the product documentation would be
.\"  appreciated but is not required.
.\"  Altered source versions must be plainly marked as such, and must not be
.\"  misrepresented as being the original software.
.\"  This notice may not be removed or altered from any source distribution.
.\"
.\" Converted to mdoc format for the OpenBSD project
.\" by Jason McIntyre <jmc@openbsd.org>
.\"
.Dd September 17, 2003
.Dt ZLIB 3
.Os
.Sh NAME
.Nm zlib
.Nd general purpose compression library
.Sh SYNOPSIS
.Fd "#include <zlib.h>"
.Pp
Basic functions
.Pp
.Ft const char *
.Fn zlibVersion "void"
.Ft int
.Fn deflateInit "z_streamp strm" "int level"
.Ft int
.Fn deflate "z_streamp strm" "int flush"
.Ft int
.Fn deflateEnd "z_streamp strm"
.Ft int
.Fn inflateInit "z_streamp strm"
.Ft int
.Fn inflate "z_streamp strm" "int flush"
.Ft int
.Fn inflateEnd "z_streamp strm"
.Pp
Advanced functions
.Pp
.Ft int
.Fn deflateInit2 "z_streamp strm" "int level" "int method" "int windowBits" "int memLevel" "int strategy"
.Ft int
.Fn deflateSetDictionary "z_streamp strm" "const Bytef *dictionary" "uInt dictLength"
.Ft int
.Fn deflateCopy "z_streamp dest" "z_streamp source"
.Ft int
.Fn deflateReset "z_streamp strm"
.Ft int
.Fn deflateParams "z_streamp strm" "int level" "int strategy"
.Ft int
.Fn inflateInit2 "z_streamp strm" "int windowBits"
.Ft int
.Fn inflateSetDictionary "z_streamp strm" "const Bytef *dictionary" "uInt dictLength"
.Ft int
.Fn inflateSync "z_streamp strm"
.Ft int
.Fn inflateReset "z_streamp strm"
.Pp
Utility functions
.Pp
.Fd typedef voidp gzFile;
.Pp
.Ft int
.Fn compress "Bytef *dest" "uLongf *destLen" "const Bytef *source" "uLong sourceLen"
.Ft int
.Fn compress2 "Bytef *dest" "uLongf *destLen" "const Bytef *source" "uLong sourceLen" "int level"
.Ft int
.Fn uncompress "Bytef *dest" "uLongf *destLen" "const Bytef *source" "uLong sourceLen"
.Ft gzFile
.Fn gzopen "const char *path" "const char *mode"
.Ft gzFile
.Fn gzdopen "int fd" "const char *mode"
.Ft int
.Fn gzsetparams "gzFile file" "int level" "int strategy"
.Ft int
.Fn gzread "gzFile file" "voidp buf" "unsigned len"
.Ft int
.Fn gzwrite "gzFile file" "const voidp buf" "unsigned len"
.Ft int
.Fn gzprintf "gzFile file" "const char *format" "..."
.Ft int
.Fn gzputs "gzFile file" "const char *s"
.Ft char *
.Fn gzgets "gzFile file" "char *buf" "int len"
.Ft int
.Fn gzputc "gzFile file" "int c"
.Ft int
.Fn gzgetc "gzFile file"
.Ft int
.Fn gzflush "gzFile file" "int flush"
.Ft z_off_t
.Fn gzseek "gzFile file" "z_off_t offset" "int whence"
.Ft int
.Fn gzrewind "gzFile file"
.Ft z_off_t
.Fn gztell "gzFile file"
.Ft int
.Fn gzeof "gzFile file"
.Ft int
.Fn gzclose "gzFile file"
.Ft const char *
.Fn gzerror "gzFile file" "int *errnum"
.Pp
Checksum functions
.Pp
.Ft uLong
.Fn adler32 "uLong adler" "const Bytef *buf" "uInt len"
.Ft uLong
.Fn crc32 "uLong crc" "const Bytef *buf" "uInt len"
.Sh DESCRIPTION
This manual page describes the
.Nm
general purpose compression library, version 1.1.4.
.Pp
The
.Nm
compression library provides in-memory compression and decompression functions,
including integrity checks of the uncompressed data.
This version of the library supports only one compression method
.Pq deflation
but other algorithms will be added later and will have the same
stream interface.
.Pp
Compression can be done in a single step if the buffers are large enough
.Pq for example if an input file is mmap'ed ,
or can be done by repeated calls of the compression function.
In the latter case, the application must provide more input
and/or consume the output
.Pq providing more output space
before each call.
.Pp
The library also supports reading and writing files in
.Xr gzip 1
.Pq .gz
format with an interface similar to that of
.Xr stdio 3 .
.Pp
The library does not install any signal handler.
The decoder checks the consistency of the compressed data,
so the library should never crash even in case of corrupted input.
.Pp
The functions within the library are divided into the following sections:
.Pp
.Bl -dash -offset indent -compact
.It
Basic functions
.It
Advanced functions
.It
Utility functions
.It
Checksum functions
.El
.Sh BASIC FUNCTIONS
.Bl -tag -width Ds
.It Xo
.Fa const char *
.Fn zlibVersion "void" ;
.Xc
.Pp
The application can compare
.Fn zlibVersion
and
.Dv ZLIB_VERSION
for consistency.
If the first character differs, the library code actually used is
not compatible with the
.Aq Pa zlib.h
header file used by the application.
This check is automatically made by
.Fn deflateInit
and
.Fn inflateInit .
.It Xo
.Fa int
.Fn deflateInit "z_streamp strm" "int level" ;
.Xc
.Pp
The
.Fn deflateInit
function initializes the internal stream state for compression.
The fields
.Fa zalloc ,
.Fa zfree ,
and
.Fa opaque
must be initialized before by the caller.
If
.Fa zalloc
and
.Fa zfree
are set to
.Dv Z_NULL ,
.Fn deflateInit
updates them to use default allocation functions.
.Pp
The compression level must be
.Dv Z_DEFAULT_COMPRESSION ,
or between 0 and 9:
1 gives best speed, 9 gives best compression, 0 gives no compression at all
(the input data is simply copied a block at a time).
.Pp
.Dv Z_DEFAULT_COMPRESSION
requests a default compromise between speed and compression
.Pq currently equivalent to level 6 .
.Pp
.Fn deflateInit
returns
.Dv Z_OK
if successful,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_STREAM_ERROR
if level is not a valid compression level,
.Dv Z_VERSION_ERROR
if the
.Nm
library version
.Pq zlib_version
is incompatible with the version assumed by the caller
.Pq ZLIB_VERSION .
.Fa msg
is set to null if there is no error message.
.Fn deflateInit
does not perform any compression: this will be done by
.Fn deflate .
.It Xo
.Fa int
.Fn deflate "z_streamp strm" "int flush" ;
.Xc
.Pp
.Fn deflate
compresses as much data as possible, and stops when the input
buffer becomes empty or the output buffer becomes full.
It may introduce some output latency
.Pq reading input without producing any output
except when forced to flush.
.Pp
The detailed semantics are as follows.
.Fn deflate
performs one or both of the following actions:
.Pp
Compress more input starting at
.Fa next_in
and update
.Fa next_in
and
.Fa avail_in
accordingly.
If not all input can be processed
(because there is not enough room in the output buffer),
.Fa next_in
and
.Fa avail_in
are updated and processing will resume at this point for the next call to
.Fn deflate .
.Pp
Provide more output starting at
.Fa next_out
and update
.Fa next_out
and
.Fa avail_out
accordingly.
This action is forced if the parameter
.Fa flush
is non-zero.
Forcing
.Fa flush
frequently degrades the compression ratio,
so this parameter should be set only when necessary
.Pq in interactive applications .
Some output may be provided even if
.Fa flush
is not set.
.Pp
Before the call to
.Fn deflate ,
the application should ensure that at least
one of the actions is possible, by providing more input and/or consuming
more output, and updating
.Fa avail_in
or
.Fa avail_out
accordingly;
.Fa avail_out
should never be zero before the call.
The application can consume the compressed output when it wants,
for example when the output buffer is full
.Pq avail_out == 0 ,
or after each call to
.Fn deflate .
If
.Fn deflate
returns
.Dv Z_OK
and with zero
.Fa avail_out ,
it must be called again after making room in the
output buffer because there might be more output pending.
.Pp
If the parameter
.Fa flush
is set to
.Dv Z_SYNC_FLUSH ,
all pending output is flushed to the output buffer and the output
is aligned on a byte boundary, so that the decompressor can get all
input data available so far.
(In particular,
.Fa avail_in
is zero after the call if enough output space
has been provided before the call.)
Flushing may degrade compression for some compression algorithms
and so it should be used only when necessary.
.Pp
If
.Fa flush
is set to
.Dv Z_FULL_FLUSH ,
all output is flushed as with
.Dv Z_SYNC_FLUSH ,
and the compression state is reset so that decompression can restart from this
point if previous compressed data has been damaged or if random access
is desired.
Using
.Dv Z_FULL_FLUSH
too often can seriously degrade the compression.
.Pp
If
.Fn deflate
returns with avail_out == 0, this function must be called again
with the same value of the flush parameter and more output space
(updated
.Fa avail_out ) ,
until the flush is complete
.Pf ( Fn deflate
returns with non-zero
.Fa avail_out ) .
.Pp
If the parameter
.Fa flush
is set to
.Dv Z_FINISH ,
pending input is processed, pending output is flushed and
.Fn deflate
returns with
.Dv Z_STREAM_END
if there was enough output space; if
.Fn deflate
returns with
.Dv Z_OK ,
this function must be called again with
.Dv Z_FINISH
and more output space
(updated
.Fa avail_out
but no more input data, until it returns with
.Dv Z_STREAM_END
or an error.
After
.Fn deflate
has returned
.Dv Z_STREAM_END ,
the only possible operations on the stream are
.Fn deflateReset
or
.Fn deflateEnd .
.Pp
.Dv Z_FINISH
can be used immediately after
.Fn deflateInit
if all the compression is to be done in a single step.
In this case,
.Fa avail_out
must be at least 0.1% larger than
.Fa avail_in
plus 12 bytes.
If
.Fn deflate
does not return
.Dv Z_STREAM_END ,
then it must be called again as described above.
.Pp
.Fn deflate
sets strm->adler to the adler32 checksum of all input read so far
(that is,
.Fa total_in
bytes).
.Pp
.Fn deflate
may update
.Fa data_type
if it can make a good guess about the input data type
.Pq Z_ASCII or Z_BINARY .
If in doubt, the data is considered binary.
This field is only for information purposes and does not affect
the compression algorithm in any manner.
.Pp
.Fn deflate
returns
.Dv Z_OK
if some progress has been made
.Pq more input processed or more output produced ,
.Dv Z_STREAM_END
if all input has been consumed and all output has been produced
(only when
.Fa flush
is set to
.Dv Z_FINISH ) ,
.Dv Z_STREAM_ERROR
if the stream state was inconsistent
(for example, if
.Fa next_in
or
.Fa next_out
was
.Dv NULL ) ,
.Dv Z_BUF_ERROR
if no progress is possible
(for example,
.Fa avail_in
or
.Fa avail_out
was zero).
.It Xo
.Fa int
.Fn deflateEnd "z_streamp strm" ;
.Xc
.Pp
All dynamically allocated data structures for this stream are freed.
This function discards any unprocessed input and does not flush any
pending output.
.Pp
.Fn deflateEnd
returns
.Dv Z_OK
if successful,
.Dv Z_STREAM_ERROR
if the stream state was inconsistent,
.Dv Z_DATA_ERROR
if the stream was freed prematurely
.Pq some input or output was discarded .
In the error case,
.Fa msg
may be set but then points to a static string
.Pq which must not be deallocated .
.It Xo
.Fa int
.Fn inflateInit "z_streamp strm" ;
.Xc
The
.Fn inflateInit
function initializes the internal stream state for decompression.
The fields
.Fa next_in ,
.Fa avail_in ,
.Fa zalloc ,
.Fa zfree ,
and
.Fa opaque
must be initialized before by the caller.
If
.Fa next_in
is not
.Dv Z_NULL
and
.Fa avail_in
is large enough
.Pq the exact value depends on the compression method ,
.Fn inflateInit
determines the compression method from the
.Nm
header and allocates all data structures accordingly;
otherwise the allocation will be deferred to the first call to
.Fn inflate .
If
.Fa zalloc
and
.Fa zfree
are set to
.Dv Z_NULL ,
.Fn inflateInit
updates them to use default allocation functions.
.Pp
.Fn inflateInit
returns
.Dv Z_OK
if successful,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_VERSION_ERROR
if the
.Nm
library version is incompatible with the version assumed by the caller.
.Fa msg
is set to null if there is no error message.
.Fn inflateInit
does not perform any decompression apart from reading the
.Nm
header if present: this will be done by
.Fn inflate .
(So
.Fa next_in
and
.Fa avail_in
may be modified,
but
.Fa next_out
and
.Fa avail_out
are unchanged.)
.It Xo
.Fa int
.Fn inflate "z_streamp strm" "int flush" ;
.Xc
.Fn inflate
decompresses as much data as possible, and stops when the input
buffer becomes empty or the output buffer becomes full.
It may introduce some output latency
.Pq reading input without producing any output
except when forced to flush.
.Pp
The detailed semantics are as follows.
.Fn inflate
performs one or both of the following actions:
.Pp
Decompress more input starting at
.Fa next_in
and update
.Fa next_in
and
.Fa avail_in
accordingly.
If not all input can be processed
(because there is not enough room in the output buffer),
.Fa next_in
is updated and processing will resume at this point for the next call to
.Fn inflate .
.Pp
Provide more output starting at
.Fa next_out
and update
.Fa next_out
and
.Fa avail_out
accordingly.
.Fn inflate
provides as much output as possible,
until there is no more input data or no more space in the output buffer
.Pq see below about the flush parameter .
.Pp
Before the call to
.Fn inflate ,
the application should ensure that at least one of the actions is possible,
by providing more input and/or consuming more output,
and updating the next_* and avail_* values accordingly.
The application can consume the uncompressed output when it wants,
for example when the output buffer is full (avail_out == 0),
or after each call to
.Fn inflate .
If
.Fn inflate
returns
.Dv Z_OK
and with zero
.Fa avail_out ,
it must be called again after making room
in the output buffer because there might be more output pending.
.Pp
If the parameter
.Fa flush
is set to
.Dv Z_SYNC_FLUSH ,
.Fn inflate
flushes as much output as possible to the output buffer.
The flushing behavior of
.Fn inflate
is not specified for values of the flush parameter other than
.Dv Z_SYNC_FLUSH
and
.Dv Z_FINISH ,
but the current implementation actually flushes as much output
as possible anyway.
.Pp
.Fn inflate
should normally be called until it returns
.Dv Z_STREAM_END
or an error.
However if all decompression is to be performed in a single step
.Pq a single call to inflate ,
the parameter
.Fa flush
should be set to
.Dv Z_FINISH .
In this case all pending input is processed and all pending output is flushed;
.Fa avail_out
must be large enough to hold all the uncompressed data.
(The size of the uncompressed data may have been saved
by the compressor for this purpose.)
The next operation on this stream must be
.Fn inflateEnd
to deallocate the decompression state.
The use of
.Dv Z_FINISH
is never required, but can be used to inform
.Fn inflate
that a faster routine may be used for the single
.Fn inflate
call.
.Pp
If a preset dictionary is needed at this point (see
.Fn inflateSetDictionary
below),
.Fn inflate
sets strm->adler to the adler32 checksum of the dictionary
chosen by the compressor and returns
.Dv Z_NEED_DICT ;
otherwise it sets strm->adler to the adler32 checksum of all output produced
so far
(that is,
.Fa total_out
bytes)
and returns
.Dv Z_OK ,
.Dv Z_STREAM_END ,
or an error code as described below.
At the end of the stream,
.Fn inflate
checks that its computed adler32 checksum is equal to that saved by the
compressor and returns
.Dv Z_STREAM_END
only if the checksum is correct.
.Pp
.Fn inflate
returns
.Dv Z_OK
if some progress has been made
.Pq more input processed or more output produced ,
.Dv Z_STREAM_END
if the end of the compressed data has been reached and all uncompressed output
has been produced,
.Dv Z_NEED_DICT
if a preset dictionary is needed at this point,
.Dv Z_DATA_ERROR
if the input data was corrupted (input stream not conforming to the
.Nm
format or incorrect adler32 checksum),
.Dv Z_STREAM_ERROR
if the stream structure was inconsistent
(for example, if
.Fa next_in
or
.Fa next_out
was
.Dv NULL ) ,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_BUF_ERROR
if no progress is possible or if there was not enough room in the output buffer
when
.Dv Z_FINISH
is used.
In the
.Dv Z_DATA_ERROR
case, the application may then call
.Fn inflateSync
to look for a good compression block.
.It Xo
.Fa int
.Fn inflateEnd "z_streamp strm" ;
.Xc
All dynamically allocated data structures for this stream are freed.
This function discards any unprocessed input and does not flush any
pending output.
.Pp
.Fn inflateEnd
returns
.Dv Z_OK
if successful, or
.Dv Z_STREAM_ERROR
if the stream state was inconsistent.
In the error case,
.Fa msg
may be set but then points to a static string
.Pq which must not be deallocated .
.El
.Sh ADVANCED FUNCTIONS
The following functions are needed only in some special applications.
.Bl -tag -width Ds
.It Xo
.Fa int
.Fn deflateInit2 "z_streamp strm" "int level" "int method" "int windowBits" "int memLevel" "int strategy" ;
.Xc
.Pp
This is another version of
.Fn deflateInit
with more compression options.
The fields
.Fa next_in ,
.Fa zalloc ,
.Fa zfree ,
and
.Fa opaque
must be initialized before by the caller.
.Pp
The
.Fa method
parameter is the compression method.
It must be
.Dv Z_DEFLATED
in this version of the library.
.Pp
The
.Fa windowBits
parameter is the base two logarithm of the window size
.Pq the size of the history buffer .
It should be in the range 8..15 for this version of the library.
Larger values of this parameter result in better compression
at the expense of memory usage.
The default value is 15 if
.Fn deflateInit
is used instead.
.Pp
The
.Fa memLevel
parameter specifies how much memory should be allocated
for the internal compression state.
memLevel=1 uses minimum memory but is slow and reduces compression ratio;
memLevel=9 uses maximum memory for optimal speed.
The default value is 8.
See
.Aq Pa zconf.h
for total memory usage as a function of
.Fa windowBits
and
.Fa memLevel .
.Pp
The
.Fa strategy
parameter is used to tune the compression algorithm.
Use the value
.Dv Z_DEFAULT_STRATEGY
for normal data;
.Dv Z_FILTERED
for data produced by a filter
.Pq or predictor ;
or
.Dv Z_HUFFMAN_ONLY
to force Huffman encoding only
.Pq no string match .
Filtered data consists mostly of small values with a
somewhat random distribution.
In this case, the compression algorithm is tuned to compress them better.
The effect of
.Dv Z_FILTERED
is to force more Huffman coding and less string matching;
it is somewhat intermediate between
.Dv Z_DEFAULT
and
.Dv Z_HUFFMAN_ONLY .
The
.Fa strategy
parameter only affects the compression ratio but not the correctness of the
compressed output, even if it is not set appropriately.
.Pp
.Fn deflateInit2
returns
.Dv Z_OK
if successful,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_STREAM_ERROR
if a parameter is invalid
.Pq such as an invalid method .
.Fa msg
is set to null if there is no error message.
.Fn deflateInit2
does not perform any compression: this will be done by
.Fn deflate .
.It Xo
.Fa int
.Fn deflateSetDictionary "z_streamp strm" "const Bytef *dictionary" "uInt dictLength" ;
.Xc
.Pp
Initializes the compression dictionary from the given byte sequence
without producing any compressed output.
This function must be called immediately after
.Fn deflateInit ,
.Fn deflateInit2 ,
or
.Fn deflateReset ,
before any call to
.Fn deflate .
The compressor and decompressor must use exactly the same dictionary
(see
.Fn inflateSetDictionary ) .
.Pp
The dictionary should consist of strings
.Pq byte sequences
that are likely to be encountered later in the data to be compressed,
with the most commonly used strings preferably put towards
the end of the dictionary.
Using a dictionary is most useful when the data to be compressed is short
and can be predicted with good accuracy;
the data can then be compressed better than with the default empty dictionary.
.Pp
Depending on the size of the compression data structures selected by
.Fn deflateInit
or
.Fn deflateInit2 ,
a part of the dictionary may in effect be discarded,
for example if the dictionary is larger than the window size in
.Fn deflate
or
.Fn deflate2 .
Thus the strings most likely to be useful should be
put at the end of the dictionary, not at the front.
.Pp
Upon return of this function, strm->adler is set to the Adler32 value
of the dictionary; the decompressor may later use this value to determine
which dictionary has been used by the compressor.
(The Adler32 value applies to the whole dictionary even if only a subset
of the dictionary is actually used by the compressor.)
.Pp
.Fn deflateSetDictionary
returns
.Dv Z_OK
if successful,
or
.Dv Z_STREAM_ERROR
if a parameter is invalid
.Pq such as NULL dictionary
or the stream state is inconsistent
(for example if
.Fn deflate
has already been called for this stream or if the compression method is bsort).
.Fn deflateSetDictionary
does not perform any compression: this will be done by
.Fn deflate .
.It Xo
.Fa int
.Fn deflateCopy "z_streamp dest" "z_streamp source" ;
.Xc
.Pp
The
.Fn deflateCopy
function sets the destination stream as a complete copy of the source stream.
.Pp
This function can be useful when several compression strategies will be
tried, for example when there are several ways of pre-processing the input
data with a filter.
The streams that will be discarded should then be freed by calling
.Fn deflateEnd .
Note that
.Fn deflateCopy
duplicates the internal compression state which can be quite large,
so this strategy is slow and can consume lots of memory.
.Pp
.Fn deflateCopy
returns
.Dv Z_OK
if successful,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_STREAM_ERROR

gnu/usr.bin/cvs/src/zlib.c
/* zlib.c --- interface to the zlib compression library
   Ian Lance Taylor <ian@cygnus.com>

   This file is part of GNU CVS.

   GNU CVS is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2, or (at your option) any
   later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* The routines in this file are the interface between the CVS
   client/server support and the zlib compression library.  */

#include <assert.h>
#include "cvs.h"
#include "buffer.h"

#if defined (SERVER_SUPPORT) || defined (CLIENT_SUPPORT)

#include "zlib.h"

/* OS/2 doesn't have EIO.  FIXME: this whole notion of turning
   a different error into EIO strikes me as pretty dubious.  */
#if !defined (EIO)
#define EIO EBADPOS
#endif

/* The compression interface is built upon the buffer data structure.
   We provide a buffer type which compresses or decompresses the data
   which passes through it.  An input buffer decompresses the data
   read from an underlying buffer, and an output buffer compresses the
   data before writing it to an underlying buffer.  */

/* This structure is the closure field of the buffer.  */

struct compress_buffer
{
    /* The underlying buffer.  */
    struct buffer *buf;
    /* The compression information.  */
    z_stream zstr;
};

static void compress_error PROTO((int, int, z_stream *, const char *));
static int compress_buffer_input PROTO((void *, char *, int, int, int *));
static int compress_buffer_output PROTO((void *, const char *, int, int *));
static int compress_buffer_flush PROTO((void *));
static int compress_buffer_block PROTO((void *, int));
static int compress_buffer_shutdown_input PROTO((void *));
static int compress_buffer_shutdown_output PROTO((void *));

/* Report an error from one of the zlib functions.  */

static void
compress_error (status, zstatus, zstr, msg)
     int status;
     int zstatus;
     z_stream *zstr;
     const char *msg;
{
    int hold_errno;
    const char *zmsg;
    char buf[100];

    hold_errno = errno;

    zmsg = zstr->msg;
    if (zmsg == NULL)
    {
        sprintf (buf, "error %d", zstatus);
	zmsg = buf;
    }

    error (status,
	   zstatus == Z_ERRNO ? hold_errno : 0,
	   "%s: %s", msg, zmsg);
}

/* Create a compression buffer.  */

struct buffer *
compress_buffer_initialize (buf, input, level, memory)
     struct buffer *buf;
     int input;
     int level;
     void (*memory) PROTO((struct buffer *));
{
    struct compress_buffer *n;
    int zstatus;

    n = (struct compress_buffer *) xmalloc (sizeof *n);
    memset (n, 0, sizeof *n);

    n->buf = buf;

    if (input)
	zstatus = inflateInit (&n->zstr);
    else
	zstatus = deflateInit (&n->zstr, level);
    if (zstatus != Z_OK)
	compress_error (1, zstatus, &n->zstr, "compression initialization");

    /* There may already be data buffered on BUF.  For an output
       buffer, this is OK, because these routines will just use the
       buffer routines to append data to the (uncompressed) data
       already on BUF.  An input buffer expects to handle a single
       buffer_data of buffered input to be uncompressed, so that is OK
       provided there is only one buffer.  At present that is all
       there ever will be; if this changes, compress_buffer_input must
       be modified to handle multiple input buffers.  */
    assert (! input || buf->data == NULL || buf->data->next == NULL);

    return buf_initialize (input ? compress_buffer_input : NULL,
			   input ? NULL : compress_buffer_output,
			   input ? NULL : compress_buffer_flush,
			   compress_buffer_block,
			   (input
			    ? compress_buffer_shutdown_input
			    : compress_buffer_shutdown_output),
			   memory,
			   n);
}

/* Input data from a compression buffer.  */

static int
compress_buffer_input (closure, data, need, size, got)
     void *closure;
     char *data;
     int need;
     int size;
     int *got;
{
    struct compress_buffer *cb = (struct compress_buffer *) closure;
    struct buffer_data *bd;

    if (cb->buf->input == NULL)
	abort ();

    /* We use a single buffer_data structure to buffer up data which
       the z_stream structure won't use yet.  We can safely store this
       on cb->buf->data, because we never call the buffer routines on
       cb->buf; we only call the buffer input routine, since that
       gives us the semantics we want.  As noted in
       compress_buffer_initialize, the buffer_data structure may
       already exist, and hold data which was already read and
       buffered before the decompression began.  */
    bd = cb->buf->data;
    if (bd == NULL)
    {
	bd = ((struct buffer_data *) malloc (sizeof (struct buffer_data)));
	if (bd == NULL)
	    return -2;
	bd->text = (char *) malloc (BUFFER_DATA_SIZE);
	if (bd->text == NULL)
	{
	    free (bd);
	    return -2;
	}
	bd->bufp = bd->text;
	bd->size = 0;
	cb->buf->data = bd;
    }

    cb->zstr.avail_out = size;
    cb->zstr.next_out = (Bytef *) data;

    while (1)
    {
	int zstatus, sofar, status, nread;

	/* First try to inflate any data we already have buffered up.
	   This is useful even if we don't have any buffered data,
	   because there may be data buffered inside the z_stream
	   structure.  */

	cb->zstr.avail_in = bd->size;
	cb->zstr.next_in = (Bytef *) bd->bufp;

	do
	{
	    zstatus = inflate (&cb->zstr, Z_NO_FLUSH);
	    if (zstatus == Z_STREAM_END)
		break;
	    if (zstatus != Z_OK && zstatus != Z_BUF_ERROR)
	    {
		compress_error (0, zstatus, &cb->zstr, "inflate");
		return EIO;
	    }
	} while (cb->zstr.avail_in > 0
		 && cb->zstr.avail_out > 0);

	bd->size = cb->zstr.avail_in;
	bd->bufp = (char *) cb->zstr.next_in;

	if (zstatus == Z_STREAM_END)
	    return -1;

	/* If we have obtained NEED bytes, then return, unless NEED is
           zero and we haven't obtained anything at all.  If NEED is
           zero, we will keep reading from the underlying buffer until
           we either can't read anything, or we have managed to
           inflate at least one byte.  */
	sofar = size - cb->zstr.avail_out;
	if (sofar > 0 && sofar >= need)
	    break;

	/* All our buffered data should have been processed at this
           point.  */
	assert (bd->size == 0);

	/* This will work well in the server, because this call will
	   do an unblocked read and fetch all the available data.  In
	   the client, this will read a single byte from the stdio
	   stream, which will cause us to call inflate once per byte.
	   It would be more efficient if we could make a call which
	   would fetch all the available bytes, and at least one byte.  */

	status = (*cb->buf->input) (cb->buf->closure, bd->text,
				    need > 0 ? 1 : 0,
				    BUFFER_DATA_SIZE, &nread);
	if (status != 0)
	    return status;

	/* If we didn't read anything, then presumably the buffer is
           in nonblocking mode, and we should just get out now with
           whatever we've inflated.  */
	if (nread == 0)
	{
	    assert (need == 0);
	    break;
	}

	bd->bufp = bd->text;
	bd->size = nread;
    }

    *got = size - cb->zstr.avail_out;

    return 0;
}

/* Output data to a compression buffer.  */

static int
compress_buffer_output (closure, data, have, wrote)
     void *closure;
     const char *data;
     int have;
     int *wrote;
{
    struct compress_buffer *cb = (struct compress_buffer *) closure;

    cb->zstr.avail_in = have;
    cb->zstr.next_in = (unsigned char *) data;

    while (cb->zstr.avail_in > 0)
    {
	char buffer[BUFFER_DATA_SIZE];
	int zstatus;

	cb->zstr.avail_out = BUFFER_DATA_SIZE;
	cb->zstr.next_out = (unsigned char *) buffer;

	zstatus = deflate (&cb->zstr, Z_NO_FLUSH);
	if (zstatus != Z_OK)
	{
	    compress_error (0, zstatus, &cb->zstr, "deflate");
	    return EIO;
	}

	if (cb->zstr.avail_out != BUFFER_DATA_SIZE)
	    buf_output (cb->buf, buffer,
			BUFFER_DATA_SIZE - cb->zstr.avail_out);
    }

    *wrote = have;

    /* We will only be here because buf_send_output was called on the
       compression buffer.  That means that we should now call
       buf_send_output on the underlying buffer.  */
    return buf_send_output (cb->buf);
}

/* Flush a compression buffer.  */

static int
compress_buffer_flush (closure)
     void *closure;
{
    struct compress_buffer *cb = (struct compress_buffer *) closure;

    cb->zstr.avail_in = 0;
    cb->zstr.next_in = NULL;

    while (1)
    {
	char buffer[BUFFER_DATA_SIZE];
	int zstatus;

	cb->zstr.avail_out = BUFFER_DATA_SIZE;
	cb->zstr.next_out = (unsigned char *) buffer;

	zstatus = deflate (&cb->zstr, Z_SYNC_FLUSH);

	/* The deflate function will return Z_BUF_ERROR if it can't do
           anything, which in this case means that all data has been
           flushed.  */
	if (zstatus == Z_BUF_ERROR)
	    break;

	if (zstatus != Z_OK)
	{
	    compress_error (0, zstatus, &cb->zstr, "deflate flush");
	    return EIO;
	}

	if (cb->zstr.avail_out != BUFFER_DATA_SIZE)
	    buf_output (cb->buf, buffer,
			BUFFER_DATA_SIZE - cb->zstr.avail_out);

	/* If the deflate function did not fill the output buffer,
           then all data has been flushed.  */
	if (cb->zstr.avail_out > 0)
	    break;
    }

    /* Now flush the underlying buffer.  Note that if the original
       call to buf_flush passed 1 for the BLOCK argument, then the
       buffer will already have been set into blocking mode, so we
       should always pass 0 here.  */
    return buf_flush (cb->buf, 0);
}

/* The block routine for a compression buffer.  */

static int
compress_buffer_block (closure, block)
     void *closure;
     int block;
{
    struct compress_buffer *cb = (struct compress_buffer *) closure;

    if (block)
	return set_block (cb->buf);
    else
	return set_nonblock (cb->buf);
}

/* Shut down an input buffer.  */

static int
compress_buffer_shutdown_input (closure)
     void *closure;
{
    struct compress_buffer *cb = (struct compress_buffer *) closure;
    int zstatus;

    /* Pick up any trailing data, such as the checksum.  */
    while (1)
    {
	int status, nread;
	char buf[100];

	status = compress_buffer_input (cb, buf, 0, sizeof buf, &nread);
	if (status == -1)
	    break;
	if (status != 0)
	    return status;
    }

    zstatus = inflateEnd (&cb->zstr);
    if (zstatus != Z_OK)
    {
	compress_error (0, zstatus, &cb->zstr, "inflateEnd");
	return EIO;
    }

    return buf_shutdown (cb->buf);
}

/* Shut down an output buffer.  */

static int
compress_buffer_shutdown_output (closure)
     void *closure;
{
    struct compress_buffer *cb = (struct compress_buffer *) closure;
    int zstatus, status;

    do
    {
	char buffer[BUFFER_DATA_SIZE];

	cb->zstr.avail_out = BUFFER_DATA_SIZE;
	cb->zstr.next_out = (unsigned char *) buffer;

	zstatus = deflate (&cb->zstr, Z_FINISH);
	if (zstatus != Z_OK && zstatus != Z_STREAM_END)
	{
	    compress_error (0, zstatus, &cb->zstr, "deflate finish");
	    return EIO;
	}

	if (cb->zstr.avail_out != BUFFER_DATA_SIZE)
	    buf_output (cb->buf, buffer,
			BUFFER_DATA_SIZE - cb->zstr.avail_out);
    } while (zstatus != Z_STREAM_END);

    zstatus = deflateEnd (&cb->zstr);
    if (zstatus != Z_OK)
    {
	compress_error (0, zstatus, &cb->zstr, "deflateEnd");
	return EIO;
    }

    status = buf_flush (cb->buf, 1);
    if (status != 0)
	return status;

    return buf_shutdown (cb->buf);
}

#endif /* defined (SERVER_SUPPORT) || defined (CLIENT_SUPPORT) */
